atomisticparsers/utils/parsers.py

```diff
@@ -0,0 +1,169 @@
+        self._info: Dict[str, Any] = {}
+        self._archive: EntryArchive = kwargs.get('archive')
+        self.thermodynamics_quantities: List[str] = ['pressure', 'temperature', 'time']
+        self.cum_max_atoms: int = 2500000
+        self.logger = get_logger(__name__)
+        self._trajectory_steps: List[int] = []
+        self._thermodynamics_steps: List[int] = []
+        self._trajectory_steps_sampled: List[int] = []
+        self._thermodynamics_steps_sampled: List[int] = []
+        self._steps: List[int] = []
+        self._steps_sampled: List[int] = []
+        for key, val in kwargs.items():
+            if hasattr(self, key):
+                setattr(self, key, val)
+
+    @property
+    def steps(self) -> List[int]:
+        '''
+        Returns the sampled steps. Steps are the set of trajectory and thermodynamics steps.
+        '''
+        if not self._steps_sampled:
+            self._steps.sort()
+            self._info['n_frames'] = len(self._steps)
+            self._steps_sampled = list(self._steps)
+            if self.frame_rate > 1:
+                self._steps_sampled = [step for n, step in enumerate(self._steps) if n % self.frame_rate == 0]
+        return self._steps_sampled
+
+    @property
+    def trajectory_steps(self) -> List[int]:
+        '''
+        Returns the sampled trajectory steps.
+        '''
+        if not self._trajectory_steps_sampled:
+            self._trajectory_steps_sampled = [
+                step for step in self._trajectory_steps if step in self.steps]
+        return self._trajectory_steps_sampled
+
+    @trajectory_steps.setter
+    def trajectory_steps(self, value: List[int]):
+        self._trajectory_steps = value
+        self._trajectory_steps.sort()
+        self._trajectory_steps_sampled = []
+        self._steps = list(set(self._trajectory_steps + self._thermodynamics_steps))
+
+    @property
+    def thermodynamics_steps(self) -> List[int]:
+        '''
+        Returns the sampled thermodynamics steps.
+        '''
+        if not self._thermodynamics_steps_sampled:
+            self._thermodynamics_steps_sampled = [
+                step for step in self._thermodynamics_steps if step in self.steps]
+        return self._thermodynamics_steps_sampled
+
+    @thermodynamics_steps.setter
+    def thermodynamics_steps(self, value: List[int]):
+        self._thermodynamics_steps = value
+        self._thermodynamics_steps.sort()
+        self._thermodynamics_steps_sampled = []
+        self._steps = list(set(self._trajectory_steps + self._thermodynamics_steps))
+
+    @property
+    def n_atoms(self) -> int:
+        return np.amax(self._info.get('n_atoms', [0]))
+
+    @n_atoms.setter
+    def n_atoms(self, value):
+        self._info['n_atoms'] = [value] if not isinstance(value, Iterable) else value
+
+    @property
+    def frame_rate(self) -> int:
+        '''
+        Returns the sampling rate of saved thermodynamics data and trajectory.
```

Comment:
Why do you say both thermodynamics and trajectory. We should only be pruning the trajectory (i.e., system) if the data is too heavy, right?

Also, maybe the comment is misleading cause it sounds like you are returning the sampling rate from the simulation, not the sampling rate for the archive. maybe clarify?

Reply:
> Also, maybe the comment is misleading cause it sounds like you are returning the sampling rate from the simulation, not the sampling rate for the archive. maybe clarify?

ah yes you are completely right here, indeed this is the sampling rate for the archive.

Reply:
> Why do you say both thermodynamics and trajectory. We should only be pruning the trajectory (i.e., system) if the data is too heavy, right
frame rate is calculated from _steps (union of thermo and traj) .

Reply:
For the frame_rate: 

Oh, ok I see now. Yeah, I don't think this makes sense in the current implementation. The frame_rate should only be calculated from the number of trajectory frames, since the criterion is based off the number of cum nr. of atoms in the trajectory.

It is conceivable that in the future there could be some higher dim quantity that is stored in calc and thus limits the frame_rate of that section, but that will have to be a separate criterion anyway, so we should just wait and address that as it comes I think.

Reply:
For the hdf5 storage: By "dump the full trajectory" do you mean the positions or the full system list of the archive or something else?

Reply:
all the positions that were parsed.

Reply:
> The frame_rate should only be calculated from the number of trajectory frames, since the criterion is based off the number of cum nr. of atoms in the trajectory.

i do not follow this, then we do not need steps and also we do not need to sample thermo_steps?

Reply:
> do not need ste

I don't really think this is necessary at the moment. There are several 3rd party modules like MDAnalysis that can read a host of output files from simulations to get the positions. If we dump to hdf5 we are just duplicating the position data that is held in the raw files (again, which are typically easily read) plus we also then have a subset again in the archive. 

Maybe in the future if someone finds that this would be easier. But it's hard for me to imagine that anyone would actually use this rather than going to the raw files. What do you think?

Reply:
This was exactly what I thought before also but if we have save them in a standard format then we can leverage also the h5web visualizer in nomad.

Reply:
> > The frame_rate should only be calculated from the number of trajectory frames, since the criterion is based off the number of cum nr. of atoms in the trajectory.
> 
> i do not follow this, then we do not need steps and also we do not need to sample thermo_steps?

I am not 100% sure that I follow the code correctly, so I don't want to tell you which variables you need or don't need.

But here is how I see the overall process:
Data in raw files:
    system_data stored every nth simulation step -- positions, atom_attributes, etc.
    calculation_data stored every mth simulation step -- thermo data, energies, etc.

Parsing strategy:
    1.  for large systems / long trajectories, prune the system_data (only) via frame_rate, which is calculated from cum_atoms = n_frames * n_atoms. n_frames here refers to the total number of steps stored (in the raw data) for system_data only.
    2. create calculation entries for all steps stored (union of stored steps for system_data and calculation_data). note - these step lists may be non-overlapping even without any pruning (nr. 1) due to the storage frequency above.

Does this help? Or which aspect is still unclear?

Reply:
1. it is cleaner now, we calculate sampling_rate only from n_traj
2. we do not sample thermo.

Reply:
> This was exactly what I thought before also but if we have save them in a standard format then we can leverage also the h5web visualizer in nomad.

I am definitely not against it if you think it is useful. Still not sure if the visualizer is useful if you only have positions in there, but maybe I don't understand the functionality.

Would you then not store the positions directly in the archive and just reference the hdf5? I imagine that the referencing could be the beneficial aspect here.

Reply:
I also have not used it but it seems you do some array manipulations. it is not efficient to store large arrays in the archive directly. We can deal with storing the full trajectory later when the need arises.

Reply:
that sounds like a reasonable plan to me